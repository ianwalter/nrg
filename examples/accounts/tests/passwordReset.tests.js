const { test } = require('@ianwalter/bff')
const app = require('../app')
const { accounts, password } = require('../seeds/01_accounts')
const { tokens } = require('../seeds/02_tokens')
const { Account, extractEmailToken } = require('@ianwalter/nrg')

const testUser = { ...accounts[1], password }
const resetVerifyUser = accounts.find(a => a.firstName === 'Reset Verify')
const readOnlyToken = tokens.find(t => t.type === 'password')

test('Password Reset • Invalid email', async t => {
  const email = 'babu_frik @example.com'
  const payload = { ...testUser, token: 'abc123', email }
  const response = await app.test('/reset-password').post(payload)
  t.expect(response.statusCode).toBe(400)
  t.expect(response.body).toMatchSnapshot()
})

test('Password Reset • Weak password', async t => {
  const payload = { ...testUser, token: 'abc123', password: 'dadudadu' }
  const response = await app.test('/reset-password').post(payload)
  t.expect(response.statusCode).toBe(400)
  t.expect(response.body).toMatchSnapshot()
})

test('Password Reset • Wrong token', async t => {
  const payload = { ...testUser, token: readOnlyToken.value }
  const response = await app.test('/reset-password').post(payload)
  t.expect(response.statusCode).toBe(400)
  t.expect(response.body).toMatchSnapshot()
})

test('Password Reset • Success', async t => {
  // Start the Forgot Password process.
  await app.test('/forgot-password').post(testUser)
  await t.asleep(1000)

  // Extract and verify the Forgot Password email and token.
  const byEmail = email => email.headers.to === testUser.email
  const { email, token } = await extractEmailToken(byEmail)
  const { action } = app.context.cfg.email.templates.passwordReset
  t.expect(email.html).toContain(action.instructions)
  t.expect(email).toMatchSnapshot({
    id: t.expect.any(String),
    messageId: t.expect.any(String),
    source: t.expect.any(String),
    date: t.expect.any(String),
    time: t.expect.any(String),
    envelope: {
      remoteAddress: t.expect.any(String),
      host: t.expect.any(String)
    },
    headers: t.expect.any(Object)
  })

  // Reset the test user's password.
  const payload = { ...testUser, token, password: 'fjioenfkj02kqwmkl606' }
  let response = await app.test('/reset-password').post(payload)
  t.expect(response.statusCode).toBe(201)
  t.expect(response.body).toMatchSnapshot({ csrfToken: t.expect.any(String) })

  // Logout.
  await app.test('/logout', response).delete()

  // Verify account data cannot be retrieved.
  response = await app.test('/account', response).get()
  t.expect(response.statusCode).toBe(401)

  // Login and verify that the new password is able to log the user in.
  response = await app.test('/login').post(payload)
  t.expect(response.statusCode).toBe(201)
  t.expect(response.body).toMatchSnapshot({ csrfToken: t.expect.any(String) })
})

test('Password Reset • Verify email through reset', async t => {
  // Start the Forgot Password process.
  await app.test('/forgot-password').post(resetVerifyUser)

  // Extract the Forgot Password token.
  await t.asleep(1000)
  const byEmail = email => email.headers.to === resetVerifyUser.email
  const { token } = await extractEmailToken(byEmail)

  // Reset the password using the token.
  const payload = { ...resetVerifyUser, token, password: 'fjioenfkj02kqwmkl60' }
  const response = await app.test('/reset-password').post(payload)
  t.expect(response.statusCode).toBe(201)

  // Verify that emailVerified is set to true in the database.
  const record = await Account.query().findById(resetVerifyUser.id)
  t.expect(record.emailVerified).toBe(true)
})
