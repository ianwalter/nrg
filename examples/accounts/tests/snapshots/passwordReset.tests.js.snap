// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Password Reset • Invalid email 1`] = `
Object {
  "feedback": Object {
    "email": Array [
      "A valid email is required.",
    ],
  },
  "message": "Validation Error",
}
`;

exports[`Password Reset • Success 1`] = `
Object {
  "date": Any<String>,
  "envelope": Object {
    "from": Object {
      "address": "support@example.com",
      "args": false,
    },
    "host": Any<String>,
    "remoteAddress": Any<String>,
    "to": Array [
      Object {
        "address": "password_reset_test@example.com",
        "args": false,
      },
    ],
  },
  "from": Array [
    Object {
      "address": "support@example.com",
      "name": "Account Example Support Team",
    },
  ],
  "headers": Any<Object>,
  "html": "
<!DOCTYPE html PUBLIC \\"-//W3C//DTD XHTML 1.0 Transitional//EN\\" \\"http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd\\">
<html xmlns=\\"http://www.w3.org/1999/xhtml\\">
<head>
  <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
  <meta http-equiv=\\"content-type\\" content=\\"text/html; charset=utf-8\\" />
  <style type=\\"text/css\\" rel=\\"stylesheet\\" media=\\"all\\">
@media only screen and (max-width: 600px) {
  .email-body_inner,
.email-footer {
    width: 100% !important;
  }
}
@media only screen and (max-width: 500px) {
  .button {
    width: 100% !important;
  }
}
</style>
</head>
<body dir=\\"ltr\\" style=\\"height: 100%; margin: 0; line-height: 1.4; background-color: #F2F4F6; color: #74787E; -webkit-text-size-adjust: none; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 100%;\\">
  <table class=\\"email-wrapper\\" width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 100%; margin: 0; padding: 0; background-color: #F2F4F6;\\" bgcolor=\\"#F2F4F6\\">
    <tr>
      <td align=\\"center\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">
        <table class=\\"email-content\\" width=\\"100%\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 100%; margin: 0; padding: 0;\\">
          <!-- Logo -->
          <tr>
            <td class=\\"email-masthead\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; padding: 25px 0; text-align: center;\\" align=\\"center\\">
              <a class=\\"email-masthead_name\\" href=\\"http://0.0.0.0:9999\\" target=\\"_blank\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; font-size: 16px; font-weight: bold; color: #2F3133; text-decoration: none; text-shadow: 0 1px 0 white;\\">
                
                  Accounts Example
                
                </a>
            </td>
          </tr>
          <!-- Email Body -->
          <tr>
            <td class=\\"email-body\\" width=\\"100%\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 100%; margin: 0; padding: 0; border-top: 1px solid #EDEFF2; border-bottom: 1px solid #EDEFF2; background-color: #FFF;\\" bgcolor=\\"#FFF\\">
              <table class=\\"email-body_inner\\" align=\\"center\\" width=\\"570\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 570px; margin: 0 auto; padding: 0;\\">
                <!-- Body content -->
                <tr>
                  <td class=\\"content-cell\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; padding: 35px;\\">
                    <h1 style=\\"margin-top: 0; color: #2F3133; font-size: 19px; font-weight: bold; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">Hi Password Reset,</h1>
                    

                    <!-- Dictionary -->
                    

                    <!-- Table -->
                    

                    <!-- Action -->
                    
                                              <p style=\\"margin-top: 0; color: #74787E; font-size: 16px; line-height: 1.5em; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">Click the button below to reset your password:</p>
                        <!--[if mso]>
                          <center>
                            <v:roundrect xmlns:v=\\"urn:schemas-microsoft-com:vml\\" xmlns:w=\\"urn:schemas-microsoft-com:office:word\\" href=\\"http://0.0.0.0:9999/reset-password?email=password_reset_test@example.com&token=<~ TOKEN WAS HERE ~>\\" style=\\"height: 45px; v-text-anchor: middle; width: 200px;\\" arcsize=\\"10%\\" stroke=\\"f\\" fillcolor=\\"#3869D4\\">
                              <w:anchorlock/>
                              <center style=\\"color: #ffffff; font-family: sans-serif; font-size: 15px;\\">
                                Reset your password
                              </center>
                            </v:roundrect>
                          </center>
                        <![endif]-->
                        <![if !mso]>
                          <table class=\\"body-action\\" align=\\"center\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 100%; margin: 30px auto; padding: 0; text-align: center;\\" width=\\"100%\\">
                            <tr>
                              <td align=\\"center\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">
                                <a href=\\"http://0.0.0.0:9999/reset-password?email=password_reset_test@example.com&token=<~ TOKEN WAS HERE ~>\\" class=\\"button\\" target=\\"_blank\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; display: inline-block; width: 200px; border-radius: 3px; color: #ffffff; font-size: 15px; line-height: 45px; text-align: center; text-decoration: none; -webkit-text-size-adjust: none; mso-hide: all; background-color: #3869D4;\\">
                                  Reset your password
                                </a>
                              </td>
                            </tr>
                          </table>
                        <![endif]>
                                          

                    <!-- Support for Gmail Go-To Actions -->
                    

                    

                    
                      <p style=\\"margin-top: 0; color: #74787E; font-size: 16px; line-height: 1.5em; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">
                        Yours truly,
                        <br>
                        Accounts Example
                      </p>
                    
                  </td>
                </tr>
              </table>
            </td>
          </tr>
          <tr>
            <td style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">
              <table class=\\"email-footer\\" align=\\"center\\" width=\\"570\\" cellpadding=\\"0\\" cellspacing=\\"0\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; width: 570px; margin: 0 auto; padding: 0; text-align: center;\\">
                <tr>
                  <td class=\\"content-cell\\" style=\\"font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; padding: 35px;\\">
                    <p class=\\"sub center\\" style=\\"margin-top: 0; line-height: 1.5em; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box; color: #AEAEAE; font-size: 12px; text-align: center;\\">
                      &copy; 2020 <a href=\\"http://0.0.0.0:9999\\" target=\\"_blank\\" style=\\"color: #3869D4; font-family: Arial, 'Helvetica Neue', Helvetica, sans-serif; -webkit-box-sizing: border-box; box-sizing: border-box;\\">Accounts Example</a>. All rights reserved.
                    </p>
                  </td>
                </tr>
              </table>
            </td>
          </tr>
        </table>
      </td>
    </tr>
  </table>
</body>
</html>
",
  "id": Any<String>,
  "messageId": Any<String>,
  "priority": "normal",
  "read": false,
  "replyTo": Array [
    Object {
      "address": "support@example.com",
      "name": "Account Example Support Team",
    },
  ],
  "source": Any<String>,
  "subject": "Accounts Example Password Reset",
  "time": Any<String>,
  "to": Array [
    Object {
      "address": "password_reset_test@example.com",
      "name": "",
    },
  ],
}
`;

exports[`Password Reset • Success 2`] = `
Object {
  "email": "password_reset_test@example.com",
  "emailVerified": true,
  "firstName": "Password Reset",
  "id": "reset",
  "lastName": "Test",
}
`;

exports[`Password Reset • Success 3`] = `
Object {
  "email": "password_reset_test@example.com",
  "emailVerified": true,
  "firstName": "Password Reset",
  "id": "reset",
  "lastName": "Test",
}
`;

exports[`Password Reset • Weak password 1`] = `
Object {
  "feedback": Object {
    "password": Array [
      "Repeats like \\"abcabcabc\\" are only slightly harder to guess than \\"abc\\"",
      "Add another word or two. Uncommon words are better.",
      "Avoid repeated words and characters",
    ],
  },
  "message": "Validation Error",
}
`;

exports[`Password Reset • Wrong token 1`] = `
Object {
  "message": "Invalid token",
}
`;
